### PR TITLE
Development

### PR DESCRIPTION
**Pull request by jaswinder9051998**
## Related Issue or bug

Info about Issue or bug.

Fixes: #[issue number that will be closed through this PR]

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Code style update (formatting, local variables).
- [ ] Breaking change (fix or feature that would cause existing functionality to stop working).
- [ ] This change requires a documentation update.

## Describe the changes you've made

addition of zoofs algos for feature selection

## How has this been tested?

- test_solver_parameter_empty_zoofs
- test_transform_zoofs
- test_objective_function_zoofs

## Is there any unusual behaviour in your code?

A clear and concise description of it.

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

## Screenshots

Original | Updated
:--------------------: | :--------------------:
**original screenshot** | **updated screenshot** |

### CODE REVIEW

**Comment by jaswinder9051998** [comment]
@tvdboom you can review the changes, all of what we had discussed is covered

File: atom/feature_engineering.py [hunk 1]
```diff
@@ -926,6 +951,80 @@ def check_y():
...
+            elif self.kwargs.get("scoring"):
+                params_for_objective_function = {
+                    "scorer": get_custom_scorer(self.kwargs["scoring"])}
+
+            else:  # default scoring method is chosen if nothing is provided
+                params_for_objective_function = {"scorer":
+                                                 (get_custom_scorer("f1")
+                                                  if y.nunique() <= 2 else
+                                                     get_custom_scorer("f1_weighted"))
+                                                 if hasattr(self._solver,
```
Comment (tvdboom) [inline atom/feature_engineering.py:1009]: Not all classifiers have a `predict_proba` method. Instead use if `self.goal == "class" else`

File: atom/feature_engineering.py [hunk 2]
```diff
@@ -529,6 +548,11 @@ class FeatureSelector(BaseEstimator, TransformerMixin, BaseTransformer, FSPlotte
             - "SFS"" Sequential Feature Selection.
             - "RFE": Recursive Feature Elimination.
             - "RFECV": RFE with cross-validated selection.
+            - "PSO" : Perform binary-particle swarm optimization for feature selection
```
Comment (tvdboom) [inline atom/feature_engineering.py:551]: What does binary mean in these defintions?

**Review by tvdboom — CHANGES_REQUESTED** [review]

File: atom/feature_engineering.py (hunk 1 above)
Comment (jaswinder9051998) [reply atom/feature_engineering.py:1009]: self.goal =='class' would work if solver is provided in atom native form like 'LGB_class' , is there any variable that tracks the task if only a model instance if provided by the user

File: atom/feature_engineering.py (hunk 2 above)
Comment (jaswinder9051998) [reply atom/feature_engineering.py:551]: the standard form of these optimizers work on continuous ( sometimes bounded ) search spaces. Feature selection being a binary task, these algos are specially modified. The 'binary' keyword is usually used to differentiate between the two forms.

File: atom/feature_engineering.py (hunk 1 above)
Comment (jaswinder9051998) [reply atom/feature_engineering.py:1009]: self.goal in FeatureSelection seems to be populated only when solver is an atom-native string format

Comment (tvdboom) [reply atom/feature_engineering.py:1009]: You are right. Use `if hasattr predict_proba or decision_function` then

**Review by tvdboom — CHANGES_REQUESTED** [review]

File: atom/feature_engineering.py (hunk 1 above)
Comment (jaswinder9051998) [reply atom/feature_engineering.py]: made the change in the latest commit

**Comment by jaswinder9051998** [comment]
All Relevant changes made

**Review by tvdboom — APPROVED** [review]

**Comment by tvdboom** [comment]
Approved. You can merge now.

**Comment by jaswinder9051998** [comment]
> Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.

I don't think I've been granted write access

**Pull request merged**